godot-macros/src/method_common.rs

```diff
@@ -0,0 +1,179 @@
+/*
+* This Source Code Form is subject to the terms of the Mozilla Public
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, You can obtain one at https://mozilla.org/MPL/2.0/.
+*/
+
+use proc_macro2::TokenStream as TokenStream2;
+use quote::quote;
+use syn::{
+    parse::{Parse, ParseStream},
+    token::Comma,
+    Ident, Result, ReturnType, Signature, Type,
+};
+
+#[derive(Copy, Clone, Eq, PartialEq)]
+pub enum ReceiverType {
+    Ref,
+    Mut,
+    Static,
+}
+
+pub struct SignatureInfo {
+    pub method_name: Ident,
+    pub receiver_type: ReceiverType,
+    pub param_idents: Vec<Ident>,
+    pub param_types: Vec<Type>,
+    pub ret_type: TokenStream2,
+}
+pub struct MethodDef {
+    pub class_name: Ident,
+    pub sig: Signature,
+}
+
+impl Parse for MethodDef {
+    fn parse(input: ParseStream) -> Result<Self> {
+        let class_name = input.parse::<Ident>()?;
+        input.parse::<Comma>()?;
+        let func = input.parse::<Signature>()?;
+
+        Ok(Self {
+            class_name,
+            sig: func,
+        })
+    }
+}
+
+pub fn wrap_with_unpacked_params(
+    class_name: &Ident,
+    signature_info: &SignatureInfo,
+) -> TokenStream2 {
+    let method_name = &signature_info.method_name;
+    let params = &signature_info.param_idents;
+
+    match signature_info.receiver_type {
+        ReceiverType::Ref => {
+            quote! {
+                |instance_ptr, params| {
+                    let ( #(#params,)* ) = params;
+
+                    let storage = unsafe { godot::private::as_storage::<#class_name>(instance_ptr) };
+                    let instance = storage.get();
+
+                    instance.#method_name( #( #params, )* )
+                }
+            }
+        }
+        ReceiverType::Mut => {
+            quote! {
+                |instance_ptr, params| {
+                    let ( #(#params,)* ) = params;
+
+                    let storage = unsafe { godot::private::as_storage::<#class_name>(instance_ptr) };
+                    let mut instance = storage.get_mut();
+
+                    instance.#method_name( #( #params, )* )
+                }
+            }
+        }
```

Comment:
These two only differentiate in `get` and `get_mut`, no? Just extract that as an ident.

Check out the `ident("...")` function we use elsewhere.